src/analysis/souffle/dataflow_graph.dl

```diff
@@ -52,14 +52,30 @@
 // to the tgt. That allows us to effectively eliminate edges that are the
 // subject of a claimNotEdge from the dataflow graph by eliminating all tags
 // from the midway point. Paths are defined in terms of resolvedEdges.
-.decl resolvedEdge(for: Principal, src: AccessPath, tgt: AccessPath)
+.decl resolvedEdge(owner: Principal, src: AccessPath, tgt: AccessPath)
 
 // A direct or transitive data flow path.
 .decl path(src: AccessPath, tgt: AccessPath)
 
+// A common way to do "forall" in Datalog is to use !exists. Unfortunately,
+// this imposes a stratification requirement. To allow IntegrityTags to
+// propagate when all predecessors have a particular IntegrityTag, instead give
+// each predecessor an arbitrary predecessor number between 0 and N, where N is
+// the number of predecessors a particular AccessPath has. Track the number of
+// predecessors an AccessPath has in numPredecssors. This allows us to
+// calculate whether all predecessors up to a particular order number have an
+// integrity tag, with the base case of 0 being just "does that particular
+// AccessPath have the IntegrityTag" and the inductive case being "does the nth
+// predecessor have the IntegrityTag and all predecessors up to N-1 have the
+// IntegrityTag". We can then ask if all predecessors up to numPredecessors
+// - 1 have the integrity tag to accomplish forall.
+.decl orderedPredecessors(ap: AccessPath, owner: Principal, src: AccessPath, orderNum: number)
```

Comment:
This is just a nit/preference so it's not strictly necessary. I'd find "dst" more intuitive than "ap" since it conveys the direction of the edge. You already have src as the other argument, so its possible to figure it out still. Similar thought for other declarations and definitions here.